daffodil-cli/src/it/scala/org/apache/daffodil/exi/TestEXIEncodeDecode.scala

```diff
@@ -0,0 +1,86 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.daffodil.saving
+
+import org.junit.Test
+import scala.xml.{Node, XML}
+
+import org.apache.daffodil.CLI.Util._
+import org.apache.daffodil.Main.ExitCode
+import org.apache.daffodil.xml.XMLUtils
+
+class TestCLIEncodeDecodeEXI {
+
+  @Test def test_3017_CLI_Encode_Decode_EXI_SA(): Unit = {
+    val schema = path("daffodil-test/src/test/resources/org/apache/daffodil/section06/namespaces/multi_base_15.dfdl.xsd")
+
+    withTempFile { tempEXI =>
+
+      // Create initial XML infoset
+      var infosetXML: Node = null
+      var encodedMD5: String = null
+      runCLI(args"parse -s $schema") { cli =>
+        cli.send("test", inputDone = true)
+        val res = cli.expectEOF
+        infosetXML = XML.loadString(res.getBefore)
+      } (ExitCode.Success)
+
+      // Encode infoset to schema aware EXI
+      runCLI(args"encodeEXI -s $schema -o $tempEXI") { cli =>
+        cli.sendLine(infosetXML.toString, inputDone = true)
+        encodedMD5 = md5sum(tempEXI)
+      } (ExitCode.Success)
+
+      // Decode EXI to XML and compare against original XML infoset
+      runCLI(args"decodeEXI -s $schema $tempEXI") { cli =>
+        val res = cli.expectEOF
+        val resultNode = XML.loadString(res.getBefore)
```

Comment:
I'm not sure how reliable getBefore is. I'm prety sure it's not guaranteed to return the full output. Instead, I'd recommend creating another tempFile that you decode to and them read that file and compare.

I htink this expectEOF + getBefore is a bit of a hack and probably should be avoided. I've removed most uses of it.

Reply:
Once you write to a tempFile, you can use commons-io to compare two temp files by calling [FileUtils.contentsEquals](https://commons.apache.org/proper/commons-io/javadocs/api-2.5/org/apache/commons/io/FileUtils.html#contentEquals(java.io.File,%20java.io.File)).